netket/logging/hdf5_log.py

```diff
@@ -0,0 +1,174 @@
+        tree_log(tree.to_dict(), root, data)  # noqa: F722
+
+    else:
+        if iter is not None:
+            tree_log(iter, f"{root}/iter", data)
+            root = f"{root}/value"
+        value = np.asarray(tree)
+        if root in data:
+            f_value = data[root]
+            f_value.resize(f_value.shape[0] + 1, axis=0)
+            f_value[-1] = value
+        else:
+            maxshape = (None,) + value.shape
+            data.create_dataset(root, data=[value], maxshape=maxshape)
+
+
+class HDF5Log(RuntimeLog):
+    """
+    HDF5 Logger, that can be passed with keyword argument `logger` to Monte
+    Carlo drivers in order to serialize the output data of the simulation.
+
+    The logger has support for scalar numbers, NumPy/JAX arrays, and netket.stats.Stats objects.
+    These are stored as individual groups within a HDF5 file, under the main group `data/`:
+
+    - scalars are stored as a group with one dataset values of shape (n_steps,) containing the logged values,
+    - arrays are stored in the same way, but with values having shape (n_steps, *array_shape),
+    - netket.stats.Stats are stored as a group containing each field (Mean, Variance, etc...) as a separate dataset.
+
+    Importantly, each group has a dataset `iters`, which tracks the iteration number of the logged quantity.
+
+    If the model state is serialized, then it is serialized as a dataset in the group `parameters/`.
+    The target of the serialization are the parameters of the variational state,
+    which are stored in a dataset `parameters/value` as a flattened array.
+
+    Data can be deserialized by calling :code:`f = h5py.File(filename, 'r')` and
+    inspecting the datasets as a dictionary, i.e. :code:`f['data/energy/Mean']`
+    """
+
+    def __init__(
+        self,
+        output_prefix: str,
+        mode: str = "write",
+        save_params: bool = True,
+        save_params_every: int = 1,
+    ):
+        """
+        Construct a HDF5 Logger.
+
+        Args:
+            output_prefix: the name of the output files before the extension
+            mode: Specify the behaviour in case the file already exists at this
+                output_prefix. Options are
+                - `[w]rite`: (default) overwrites file if it already exists;
+                - `[x]` or `fail`: fails if file already exists;
+            save_params: bool flag indicating whether variables of the variational state
+                should be serialized at some interval
+            save_params_every: every how many iterations should machine parameters be
+                flushed to file
+        """
+        import h5py
+
+        super().__init__()
+
+        if not ((mode == "write") or (mode == "append") or (mode == "fail")):
+            raise ValueError(
+                "Mode not recognized: should be one of `[w]rite`, `[a]ppend` or"
+                "`[x]`(fail)."
+            )
+        mode = _mode_shorthands[mode]
+
```

Comment:
I should add this to the other loggers as well, it always gets me. So since we're here... let's make it work also in the case that one specifies the `.hdf5` as well.
```suggestion

    if output_prefix.endswith(".hdf5"):
        output_prefix = output_prefix[:-5]

```